server-shared/src/main/java/rikka/shizuku/server/UserService.java

```diff
@@ -54,25 +54,35 @@ public static Pair<IBinder, String> create(String[] args) {
         IBinder service;
 
         try {
-            Context systemContext = ActivityThread.systemMain().getSystemContext();
+            ActivityThread activityThread = ActivityThread.systemMain();
+            Context systemContext = activityThread.getSystemContext();
 
-            DdmHandleAppName.setAppName(name != null ? name : pkg + ":user_service", 0);
+            DdmHandleAppName.setAppName(name != null ? name : pkg + ":user_service", userId);
 
             //noinspection InstantiationOfUtilityClass
             UserHandle userHandle = Refine.unsafeCast(
                     Build.VERSION.SDK_INT >= Build.VERSION_CODES.N
                             ? UserHandleHidden.of(userId)
                             : new UserHandleHidden(userId));
             Context context = Refine.<ContextHidden>unsafeCast(systemContext).createPackageContextAsUser(pkg, Context.CONTEXT_INCLUDE_CODE | Context.CONTEXT_IGNORE_SECURITY, userHandle);
-            ClassLoader classLoader = context.getClassLoader();
+            Field mPackageInfo = context.getClass().getDeclaredField("mPackageInfo");
+            mPackageInfo.setAccessible(true);
+            Object loadedApk = mPackageInfo.get(context);
+            Method makeApplication = loadedApk.getClass().getDeclaredMethod("makeApplication", boolean.class, Instrumentation.class);
+            Application application = (Application) makeApplication.invoke(loadedApk, true, null);
+            Field mInitialApplication = activityThread.getClass().getDeclaredField("mInitialApplication");
+            mInitialApplication.setAccessible(true);
+            mInitialApplication.set(activityThread, application);
```

Comment:
Looks to me like this warrants an addition to https://github.com/RikkaApps/Shizuku-API/tree/master/demo-hidden-api-stub instead of doing reflection directly here, wdyt?

Reply:
I see, probably wouldn't work because using heavy usage of private fields, nvm my comment then